packages/dd-trace/src/config-env-sources.js

```diff
@@ -0,0 +1,112 @@
+'use strict'
+
+const { isInServerlessEnvironment } = require('./serverless')
+const { getValueFromEnvSource, getEnvironmentVariable } = require('./config-helper')
+
+/**
+ * Internal helper that centralizes configuration coming from stable config files and environment variables.
+ *
+ * Responsibilities:
+ * - Load and parse the local and fleet stable configuration files (via `StableConfig`).
+ * - Expose the stable config maps (`localStableConfig`, `fleetStableConfig`) and any warnings
+ *   `stableConfigWarnings`.
+ * - Provide `getResolvedEnv(name)` to resolve a configuration value from:
+ *     1. Fleet stable config
+ *     2. Environment variables (and their aliases)
+ *     3. Local stable config
+ *
+ * Usage patterns:
+ * - Most modules that need a single resolved value should call `getResolvedEnv(name)`.
+ * - The main `Config` class reads `localStableConfig` / `fleetStableConfig` directly so it can:
+ *     - Preserve per-source origins (fleet/local/env) for telemetry reporting.
+ *     - Merge sources in its own priority order.
+ * - Tests can reset the singleton with `resetConfigEnvSources()` to force re-loading after
+ *   modifying stable config files or environment variables.
+ */
+class ConfigEnvSources {
+  constructor () {
+    const isServerless = isInServerlessEnvironment()
+
+    let localStableConfig = {}
+    let fleetStableConfig = {}
+    let stableConfigWarnings = []
+
+    if (!isServerless) {
+      const result = this.#loadStableConfig()
+      if (result) {
+        localStableConfig = result.localEntries
+        fleetStableConfig = result.fleetEntries
+        stableConfigWarnings = result.warnings
+      }
+    }
+
+    // Expose raw stable config on the instance
+    this.localStableConfig = localStableConfig
+    this.fleetStableConfig = fleetStableConfig
+    this.stableConfigWarnings = stableConfigWarnings
+  }
+
+  #loadStableConfig () {
+    try {
+      const StableConfig = require('./config_stable')
+      const instance = new StableConfig()
+      return {
+        localEntries: instance.localEntries ?? {},
+        fleetEntries: instance.fleetEntries ?? {},
+        warnings: instance.warnings ?? []
+      }
+    } catch {
+      // Stable config is optional, continue without it
+      return null
+    }
+  }
+}
+
+function createConfigEnvSources () {
+  return new ConfigEnvSources()
+}
+
+let configEnvSourcesInstance = null
+
+function getConfigEnvSources () {
+  if (!configEnvSourcesInstance) {
+    configEnvSourcesInstance = new ConfigEnvSources()
+  }
+  return configEnvSourcesInstance
+}
+
+/**
+ * Reset the singleton instance (useful for testing)
+ */
+function resetConfigEnvSources () {
+  configEnvSourcesInstance = null
+}
```

Comment:
I think we can just use proxyquire to achieve the same instead of writing dedicated production code to make things testable :)